### PR TITLE
Jd/add forward macro

### PR DESCRIPTION
**Pull request by jd-lara**
This PR adds the @forward macro capability to IS. The code can't be tested here because IS won't export any method so the tests for this will be in PSY.

### CODE REVIEW

File: src/components.jl [hunk 1]
```diff
@@ -215,7 +215,7 @@ function get_components_by_name(
     components::Components,
     name::AbstractString,
 )::Vector{T} where {T <: InfrastructureSystemsType}
-    if !isabstracttype(T)
+    if isconcretetype(T)
         throw(ArgumentError("get_components_by_name only supports abstract types: $T"))
```
Comment (daniel-thom) [inline src/components.jl:219]: Can you change the error string?  ```does not support concrete types```

**Comment by jd-lara** [comment]
bump @kdheepak for this review. I will leave it open for another 24 hours for comments.

File: src/component.jl [hunk 2]
```diff
@@ -390,6 +390,47 @@ function check_contiguous_forecasts(
...
+
+# Arguments
+- `src::InfrastructureSystemsType`: Source component
+- `dst::InfrastructureSystemsType`: Destination component
+- `label_mapping::Dict = nothing`: Optionally map src labels to different dst labels.
+  If provided and src has a forecast with a label not present in label_mapping, that
+  forecast will not copied. If label_mapping is nothing then all forecasts will be copied
+  with src's labels.
+"""
+function copy_forecasts!(
```
Comment (daniel-thom) [inline src/component.jl:405]: Why are these changes showing up?  They are in master.

File: src/utils/utils.jl [hunk 3]
```diff
@@ -309,3 +309,89 @@ macro scoped_enum(T, args...)
...
+function forward(sender::Tuple{Type, Symbol}, receiver::Type)
+    code = Vector{String}()
+    active_methods = getfield.(InteractiveUtils.methodswith(sender[1]), :name)
+    for m in InteractiveUtils.methodswith(receiver)
+        m.name ∈ active_methods && continue
+        if startswith(string(m.name), "get_") && m.nargs == 2
+            # forwarding works for functions with 1 argument and starts with `get_`
+            append!(code, forward(sender, receiver, m))
+        end
+        if startswith(string(m.name), "set_") && m.nargs == 3
```
Comment (daniel-thom) [inline src/utils/utils.jl:381]: elseif

File: src/component.jl (hunk 2 above)
Comment (jd-lara) [reply src/component.jl:405]: I don't know, I think that when I rebased they showed up here.

**Pull request merged**